Card Directive Expansion (EM-1320: Small Biz Add-Ons Section)

toastercup opened the pull request:
This PR features the buildout of the `card-blocks` component, which was yanked from `employer` and refactored. Specifically, the following bits and bobs have changed:
* Fixes issue where text/buttons would scale independently of the block itself
* Bourbon's `transform` prefixer now wraps around our use of `scale`
* `scale3d` -> `scale`
* Lots of unnecessary styles removed or unified
* `card-block` can take in a `$include_strip` parameter to toggle the strip along the top of the card
* ...and more!

arelia commented on sass/directives/_card_blocks.scss:
```diff
@@ -0,0 +1,66 @@
+      @include transform(scale(1.06));
+      @include transition(transform 550ms ease-out);
+
+      &:before {
```
':before' 'background-color' can also be included in `if` block. there's nothing to change unless there's a strip

arelia commented on sass/directives/_card_blocks.scss:
```diff
@@ -0,0 +1,66 @@
+  }
+}
+
+@mixin card-block__logo {
```
I don't think this makes sense as a mixin by itself. the name of the mixin doesn't imply what it does, and it only does one thing which could be handled as a variable. what might be better is 

```scss
@mixin card-block__logo {
  img {
    height: 80px; // this could be a parameter on this mixin, and/or a variable used within the mixin (i'm leaning towards variable, since we might not have cards with varying logo sizes anytime soon, so we can just set a variable for a standard size)
    margin-bottom: $base-spacing-small; // moving this from line 15 above, since we only need this margin if there's a logo present
  }
}
```
this would get included when using the `card--tile` mixin when there's a logo nested in that block, i.e.

```scss
.selector {
  @include card--tile;
  @include card-block__logo; // or rename to card--tile__logo or card__logo
}
```

toastercup replied:
I like this - my only problem with it is that it would apply these styles to any image within the card. I've implemented it for now, but we may want to revisit it one day. I'm still of the pursuasion that we'd want to apply this mixin directly to a semantic class just for the image tag, for maximum control.

However: the original use case I created for this mixin is a bit silly. There's no reason to create a mixin that takes in a single parameter (height) and applies a single style - that's reinventing the wheel. To this end, I would remove `height` from this mixin and apply it directly to the semantic class (and keep the mixin to apply the `margin-bottom`). Since we won't be using a semantic class for the image in this particular implementation, however, I'll continue taking in `height` as a parameter. Thanks!

arelia commented on sass/directives/_card_blocks.scss:
```diff
@@ -0,0 +1,66 @@
+}
+
+@mixin card-block__button {
+  text-transform: uppercase;
```
`text-transform: uppercase;` probably comes from before the button updates. The button will have this by default now so this can be removed

arelia commented on sass/directives/_card_blocks.scss:
```diff
@@ -0,0 +1,66 @@
+@mixin card-block__button {
+  text-transform: uppercase;
+  margin-top: $base-spacing-medium; // TODO standardize new button top margins
+  margin-bottom: $base-spacing-small;
```
This adds extra space between the bottom of the button and the bottom of the containing block. That should be handled by the containing blocks padding, so this can be removed and padding adjusted if Susan requested it

The padding comes from component-box, it's the standard padding for all boxes: https://github.com/cb-talent-development/employer-style-base/blob/bugfix/fieldset-large-input-button-alignment/sass/directives/_boxes.scss#L3

arelia commented on sass/directives/_card_blocks.scss:
```diff
@@ -0,0 +1,66 @@
+  @include flex-shrink(0);
+  @include flex-basis(auto);
+
+  color: $body-text-color;
```
this `color` can be removed, the `h3` gets this color by default and the `p` gets a different color

arelia commented on sass/directives/_card_blocks.scss:
```diff
@@ -0,0 +1,66 @@
+  margin-bottom: $base-spacing-small;
+}
+
+@mixin card-block__text {
```
I'm good with these mixins that define what the content looks like separately from how the box looks 👍

arelia commented on sass/directives/_card_blocks.scss:
```diff
@@ -0,0 +1,66 @@
+  }
+}
+
+@mixin card-block__text--narrow { // TODO is there another solution to achieve design + not having tweaks for two of the blocks?
```
per here: https://github.com/cbdr/employer/pull/850/files#r117911336 this gets removed/consolidated with the text mixin

toastercup commented on sass/directives/_boxes.scss:
```diff
@@ -51,12 +51,19 @@
   text-decoration: none;
+  font-weight: $font-weight-normal;
+
+  &:hover {
```
@arelia unfortunately, there's no way to re-introduce the active/focus state styles without having them be triggered in unexpected ways. We'll have to explore alternatives in a defect. In the meantime, I will re-introduce the states and note in the Trello card, PR and JIRA ticket that their will be unexpected behavior on this page. Thanks for bringing this up!

toastercup commented on sass/directives/_boxes.scss:
```diff
@@ -51,12 +51,19 @@
   &:active,
-  &:focus {
+  &:focus,
+  a:hover,
```
This change stemmed from my struggle getting Magnific popups to work with our card structure - it's definitely no longer necessary. Good catch!

toastercup commented:
@arelia I've addressed your feedback, and this is now ready for re-review. I've 'graduated' the `card-tile` mixins to their own file, as they felt distinct from `box`es to me. Thanks!

arelia commented on sass/directives/_cards.scss:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  @media only screen and (min-width: $small-screen-max) { // TODO turning off the fanciness on mobile but should there still be some sort of state for usability?
+    &:hover {
```
this `transform` and `transition` should also be wrapped in `@if $scale`

arelia commented on sass/directives/_cards.scss:
```diff
@@ -0,0 +1,113 @@
+}
+
+@mixin card--tile__logo($logo_height: 80px) {
+  img {
```
you mentioned this would affect any other image inside the box. true. maybe this should be applied directly to a semantic class on logo images

toastercup commented:
@arelia last bits of feedback addressed - thanks!
